File: src/main/java/com/taw/polybank/service/PaymentService.java
package com.taw.polybank.service;

import com.taw.polybank.dao.PaymentRepository;
import com.taw.polybank.dto.BenficiaryDTO;
import com.taw.polybank.dto.CurrencyExchangeDTO;
import com.taw.polybank.dto.PaymentDTO;
import com.taw.polybank.entity.PaymentEntity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class PaymentService {
    @Autowired
    protected PaymentRepository paymentRepository;

    public void save(PaymentDTO paymentDTO, BeneficiaryService beneficiaryService, CurrencyExchangeService currencyExchangeService, BadgeService badgeService) {
        PaymentEntity payment = this.toEntity(paymentDTO, beneficiaryService, currencyExchangeService, badgeService);
        paymentRepository.save(payment);
        paymentDTO.setId(payment.getId());
    }

    public PaymentEntity toEntity(PaymentDTO paymentDTO, BeneficiaryService beneficiaryService, CurrencyExchangeService currencyExchangeService, BadgeService badgeService) {
        PaymentEntity payment = paymentRepository.findById(paymentDTO.getId()).orElse(new PaymentEntity());
        payment.setId(paymentDTO.getId());
        payment.setAmount(paymentDTO.getAmount());
        payment.setBenficiaryByBenficiaryId(beneficiaryService.toEntity(paymentDTO.getBenficiaryByBenficiaryId()));
        payment.setCurrencyExchangeByCurrencyExchangeId(currencyExchangeService.toEntity(paymentDTO.getCurrencyExchangeByCurrencyExchangeId(), badgeService));
        return payment;
    }
}

File: src/main/java/com/taw/polybank/service/PaymentService.kt
package com.taw.polybank.service

import com.taw.polybank.dao.PaymentRepository
import com.taw.polybank.dto.PaymentDTO
import com.taw.polybank.entity.PaymentEntity
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

/**
 * @author Illya Rozumovskyy
 */
@Service
class PaymentService(
  private val paymentRepository: PaymentRepository,
) {
  fun save(
    paymentDTO: PaymentDTO,
    beneficiaryService: BeneficiaryService,
    currencyExchangeService: CurrencyExchangeService,
    badgeService: BadgeService,
  ) {
    val payment =
      this.toEntity(paymentDTO, beneficiaryService, currencyExchangeService, badgeService)
    paymentRepository.save<PaymentEntity>(payment)
    paymentDTO.setId(payment.getId())
  }

  fun toEntity(
    paymentDTO: PaymentDTO,
    beneficiaryService: BeneficiaryService,
    currencyExchangeService: CurrencyExchangeService,
    badgeService: BadgeService,
  ): PaymentEntity {
    val payment =
      paymentRepository.findById(paymentDTO.getId()).orElse(PaymentEntity())
    payment.setId(paymentDTO.getId())
    payment.setAmount(paymentDTO.getAmount())
    payment.setBenficiaryByBenficiaryId(
      beneficiaryService.toEntity(paymentDTO.getBenficiaryByBenficiaryId()),
    )
    val currencyExchangeEntity =
      if (paymentDTO.getCurrencyExchangeByCurrencyExchangeId() == null) {
        null
      } else {
        currencyExchangeService.toEntity(
          paymentDTO.getCurrencyExchangeByCurrencyExchangeId(),
          badgeService,
        )
      }
    payment.setCurrencyExchangeByCurrencyExchangeId(currencyExchangeEntity)
    return payment
  }
}
